pkg/controllers/dashboard/plugin/fscache.go

```diff
@@ -0,0 +1,279 @@
+func (c FSCache) SyncWithIndex(index *SafeIndex, fsCacheFiles []string) error {
+	for _, file := range fsCacheFiles {
+		logrus.Debugf("syncing index with filesystem cache")
+		chartName, chartVersion, err := getChartNameAndVersion(file)
+		if err != nil {
+			return err
+		}
+		_, ok := index.Entries[chartName]
+		if !ok || index.Entries[chartName].Version != chartVersion {
+			err := c.Delete(chartName, chartVersion)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// Save takes in data and a path to save it in the filesystem cache
+func (c FSCache) Save(data []byte, path string) error {
+	logrus.Debugf("creating file [%s]", path)
+	if err := os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+		return err
+	}
+	out, err := os.OpenFile(path, syscall.O_RDWR|syscall.O_CREAT|syscall.O_TRUNC, 0600)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	out.Write(data)
+
+	return nil
+}
+
+// Delete takes in a plugin's name and version, and deletes its entry in the filesystem cache
+func (c FSCache) Delete(name, version string) error {
+	p, err := filepathsecure.SecureJoin(FSCacheRootDir, name)
+	if err != nil {
+		return err
+	}
+	err = osRemoveAll(p)
+	if err != nil {
+		err = fmt.Errorf("failed to delete entry [Name: %s Version: %s] from filesystem cache: %w", name, version, err)
+		return err
+	}
+	logrus.Debugf("deleted plugin entry from cache [Name: %s Version: %s]", name, version)
+
+	return nil
+}
+
+// isCached takes in the name and version of a plugin and returns true if
+// it is cached (entry exists and files were fetched), returns false otherwise
+func (c FSCache) isCached(name, version string) (bool, error) {
+	path, err := filepathsecure.SecureJoin(FSCacheRootDir, filepath.Join(name, version))
+	if err != nil {
+		return false, err
+	}
+	_, err = osStat(path)
+	if err != nil {
+		if errors.Is(err, os.ErrNotExist) {
+			return false, nil
+		}
+		return false, err
+	}
+	isEmpty, err := isDirEmpty(path)
+	if err != nil {
+		return false, err
+	}
+	if !isEmpty {
+		return true, nil
+	}
+	return false, nil
+}
+
+// getChartNameAndVersion receives a filepath and returns the chart name and chart version.
+// The path has to follow /{root}/{chartName}/{chartVersion}/*
+func getChartNameAndVersion(file string) (string, string, error) {
+	root, err := filepath.Abs(FSCacheRootDir)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get absolute root path: %w", err)
+	}
+	filePath, err := filepath.Abs(file)
+	if err != nil {
+		return "", "", fmt.Errorf("unable to get absolute file path: %w", err)
+	}
+
+	// file has to be rooted at FSCacheRootDir
+	if strings.HasPrefix(filePath, root) {
+		p, _ := strings.CutPrefix(filePath, root)
+		s := strings.Split(p, string(os.PathSeparator))
+		if len(s) < 3 {
+			return "", "", fmt.Errorf("file path is not valid. Path provided: %s", file)
+		}
+		chartName := s[1]
+		chartVersion := s[2]
+		if _, err := semver.NewVersion(chartVersion); err != nil {
```

Comment:
I think we shouldn't check for semver here. The right place to check is when this is being added or I am okay with not checking at all. But I think this is not the right place to make this validation.

Reply:
True, it's a duplicated validation